internal/provisioner/registry/controllers/bundle_controller.go

```diff
@@ -0,0 +1,285 @@
+		if err != nil {
+			return ctrl.Result{}, updateStatusUnpackFailing(&u, fmt.Errorf("convert registry+v1 bundle to plain+v0 bundle: %w", err))
+		}
+
+		objects, err := getObjects(plainFS)
+		if err != nil {
+			return ctrl.Result{}, updateStatusUnpackFailing(&u, fmt.Errorf("get objects from bundle manifests: %w", err))
+		}
+		if len(objects) == 0 {
+			return ctrl.Result{}, updateStatusUnpackFailing(&u, errors.New("invalid bundle: found zero objects: "+
+				"plain+v0 bundles are required to contain at least one object"))
+		}
+
+		if err := r.Storage.Store(ctx, bundle, plainFS); err != nil {
+			return ctrl.Result{}, updateStatusUnpackFailing(&u, fmt.Errorf("persist bundle objects: %w", err))
+		}
+
+		contentURL, err := r.Storage.URLFor(ctx, bundle)
+		if err != nil {
+			return ctrl.Result{}, updateStatusUnpackFailing(&u, fmt.Errorf("get content URL: %w", err))
+		}
+
+		updateStatusUnpacked(&u, unpackResult, contentURL)
+		return ctrl.Result{}, nil
+	default:
+		return ctrl.Result{}, updateStatusUnpackFailing(&u, fmt.Errorf("unknown unpack state %q: %w", unpackResult.State, err))
+	}
+}
+
+func updateStatusUnpackPending(u *updater.Updater, result *source.Result) {
+	u.UpdateStatus(
+		updater.EnsureResolvedSource(nil),
+		updater.EnsureContentURL(""),
+		updater.SetPhase(rukpakv1alpha1.PhasePending),
+		updater.EnsureCondition(metav1.Condition{
+			Type:    rukpakv1alpha1.TypeUnpacked,
+			Status:  metav1.ConditionFalse,
+			Reason:  rukpakv1alpha1.ReasonUnpackPending,
+			Message: result.Message,
+		}),
+	)
+}
+
+func updateStatusUnpacking(u *updater.Updater, result *source.Result) {
+	u.UpdateStatus(
+		updater.EnsureResolvedSource(nil),
+		updater.EnsureContentURL(""),
+		updater.SetPhase(rukpakv1alpha1.PhaseUnpacking),
+		updater.EnsureCondition(metav1.Condition{
+			Type:    rukpakv1alpha1.TypeUnpacked,
+			Status:  metav1.ConditionFalse,
+			Reason:  rukpakv1alpha1.ReasonUnpacking,
+			Message: result.Message,
+		}),
+	)
+}
+
+func updateStatusUnpacked(u *updater.Updater, result *source.Result, contentURL string) {
+	u.UpdateStatus(
+		updater.EnsureResolvedSource(result.ResolvedSource),
+		updater.EnsureContentURL(contentURL),
+		updater.SetPhase(rukpakv1alpha1.PhaseUnpacked),
+		updater.EnsureCondition(metav1.Condition{
+			Type:    rukpakv1alpha1.TypeUnpacked,
+			Status:  metav1.ConditionTrue,
+			Reason:  rukpakv1alpha1.ReasonUnpackSuccessful,
+			Message: result.Message,
+		}),
+	)
+}
+
+func updateStatusUnpackFailing(u *updater.Updater, err error) error {
+	u.UpdateStatus(
+		updater.EnsureResolvedSource(nil),
+		updater.EnsureContentURL(""),
+		updater.SetPhase(rukpakv1alpha1.PhaseFailing),
+		updater.EnsureCondition(metav1.Condition{
+			Type:    rukpakv1alpha1.TypeUnpacked,
+			Status:  metav1.ConditionFalse,
+			Reason:  rukpakv1alpha1.ReasonUnpackFailed,
+			Message: err.Error(),
+		}),
+	)
+	return err
+}
+
+func getObjects(bundleFS fs.FS) ([]client.Object, error) {
```

Comment:
maybe we could provide a closure here for how to handle building up the output array of client.Object(s) here if we're using this logic in multiple provisioner implementations?